File: iosApp/iosApp/Modules/Account/AccountInteractor.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

protocol AccountInteractor: AnyObject {
    func address() -> String
    var network: Network { get }
    var loadingTransactions: Bool { get }
    func currency() -> Currency
    func metadata() -> CurrencyMetadata?
    func market() -> CurrencyMarketData?
    func candles() -> [Candle]?
    func cryptoBalance() -> BigInt
    func fiatBalance() -> Double
    func transactions() -> [AccountInteractorTransaction]
    func fetchTransactions(_ handler: @escaping ([AccountInteractorTransaction]) -> ())
}

struct AccountInteractorTransaction {
    let date: Date?
    let blockNumber: String
    let address: String
    let amount: String
    let isReceive: Bool
}

final class DefaultAccountInteractor {
    private let wallet: Wallet
    private (set) var network: Network
    private let _currency: Currency
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    private let walletService: WalletService
    private let transactionService: EtherscanService

    private(set) var loadingTransactions: Bool = false

    init(
        wallet: Wallet,
        currency: Currency,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService,
        walletService: WalletService,
        transactionService: EtherscanService
    ) {
        self.wallet = wallet
        self.network = wallet.network() ?? Network.ethereum()
        self._currency = currency
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
        self.walletService = walletService
        self.transactionService = transactionService
    }
}

extension DefaultAccountInteractor: AccountInteractor {

    func address() -> String {
        // TODO: Review this if correct Annon, should not this be currency.address and if nil this?
        walletService.address(network: network) ?? ""
    }

    func currency() -> Currency {
        self._currency
    }

    func metadata() -> CurrencyMetadata? {
        currencyStoreService.metadata(currency: _currency)
    }

    func market() -> CurrencyMarketData? {
        currencyStoreService.marketData(currency: _currency)
    }

    func candles() -> [Candle]? {
        currencyStoreService.candles(currency: _currency)
    }

    func cryptoBalance() -> BigInt {
        walletService.balance(network: network, currency: _currency)
    }

    func fiatBalance() -> Double {
        let price = currencyStoreService.marketData(currency: _currency)?
            .currentPrice?
            .doubleValue ?? 0
        return web3lib.CurrencyFormatter.Companion().crypto(
            amount: cryptoBalance(),
            decimals: _currency.decimals(),
            mul: price
        )
    }

    func transactions() -> [AccountInteractorTransaction] {
        guard let address = walletService.address(network: network) else {
            return []
        }
        guard currency().type != .erc20 else {
            let logs = walletService.transferLogs(currency: currency(), network: network)
            return toTransactions(from: logs)
        }
        return toTransactions(from:
            transactionService.cachedTransactionHistory(
                for: address,
                network: network,
                nonce: walletService.transactionCount(network: network).toDecimalString()
            )
        )
    }

    func fetchTransactions(_ handler: @escaping ([AccountInteractorTransaction]) -> ()) {
        guard let address = walletService.address(network: network) else { return }

        loadingTransactions = true

        guard currency().type != .erc20 else {
            return walletService.fetchTransferLogs(
                currency: currency(),
                network: network,
                completionHandler: { logs, error in
                    DispatchQueue.main.async { [weak self] in
                        guard let self = self else { return }
                        if let error = error { print(error) }
                        self.loadingTransactions = false
                        handler(self.toTransactions(from: logs ?? []))
                    }
                }
            )
        }

        transactionService.fetchTransactionHistory(for: address, network: network) { result in
            DispatchQueue.main.async { [weak self] in
                guard let self = self else { return }
                self.loadingTransactions = false
                switch result {
                case let .success(transactions):
                    handler(self.toTransactions(from: transactions))
                case let .failure(error):
                    print(error)
                    handler([])
                }
            }
        }
    }

    func toTransactions(
        from transactions: [EtherscanTransaction]
    ) -> [AccountInteractorTransaction] {
        transactions
            .filter { $0.value != "0" }
            .map {
                let isReceive = $0.to == self.address()
                return .init(
                    date: Date(timeIntervalSince1970: (try? $0.timeStamp.double()) ?? 0),
                    blockNumber: $0.blockNumber,
                    address: isReceive ? $0.from : $0.to,
                    amount: Formatter.currency.string(
                        BigInt.fromString($0.value, decimals: 0),
                        currency: currency(),
                        style: .long(minDecimals: 8)
                    ),
                    isReceive: isReceive
                )
        }
    }

    func toTransactions(
        from transactions: [Log]
    ) -> [AccountInteractorTransaction] {
        transactions.map {
            guard let topic1 = ($0.topics?[1] as? Topic.TopicValue)?.value,
                  let topic2 = ($0.topics?[2] as? Topic.TopicValue)?.value else {
                return nil
            }
            let from: Address.HexString = abiDecodeAddress(topic1)
            let to: Address.HexString = abiDecodeAddress(topic2)
            let amount: BigInt = abiDecodeBigInt($0.data)
            let isReceive = to.hexString == self.address()
            return .init(
                date: nil,
                blockNumber: $0.blockNumber.toDecimalString(),
                address: isReceive ? from.hexString : to.hexString,
                amount: Formatter.currency.string(
                    amount,
                    currency: currency(),
                    style: .long(minDecimals: 8)
                ),
                isReceive: isReceive
            )
        }.compactMap { $0 }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/account/AccountInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.account

import com.sonsofcrypto.web3lib.formatters.Formater
import com.sonsofcrypto.web3lib.provider.model.Log
import com.sonsofcrypto.web3lib.provider.model.Topic
import com.sonsofcrypto.web3lib.services.coinGecko.model.Candle
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyMarketData
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyMetadata
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Currency.Type.ERC20
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.types.BigInt
import com.sonsofcrypto.web3lib.legacy.abiDecodeAddress
import com.sonsofcrypto.web3lib.legacy.abiDecodeBigInt
import com.sonsofcrypto.web3lib.legacy.BlockNumberToTimestampHelper
import com.sonsofcrypto.web3walletcore.services.etherScan.EtherScanService
import com.sonsofcrypto.web3walletcore.services.etherScan.EtherScanTransaction

data class AccountTransaction(
    val date: Int?,
    val blockNumber: String,
    val from: String,
    val to: String,
    val amount: BigInt,
    val txHash: String,
)

interface AccountInteractor {
    fun address(network: Network): String
    fun metadata(currency: Currency): CurrencyMetadata?
    fun market(currency: Currency): CurrencyMarketData?
    fun candles(currency: Currency): List<Candle>?
    fun cryptoBalance(network: Network, currency: Currency): BigInt
    fun fiatBalance(network: Network, currency: Currency): Double
    fun fiatPrice(amount: BigInt, currency: Currency): Double
    fun transactions(network: Network, currency: Currency): List<AccountTransaction>
    @Throws(Throwable::class)
    suspend fun fetchTransactions(network: Network, currency: Currency)
    fun isVoidSigner(): Boolean
}

class DefaultAccountInteractor(
    private val currencyStoreService: CurrencyStoreService,
    private val walletService: WalletService,
    private val etherScanService: EtherScanService,
): AccountInteractor {

    override fun address(network: Network): String = walletService.address(network) ?: ""

    override fun metadata(currency: Currency): CurrencyMetadata? =
        currencyStoreService.metadata(currency)

    override fun market(currency: Currency): CurrencyMarketData? =
        currencyStoreService.marketData(currency)

    override fun candles(currency: Currency): List<Candle>? =
        currencyStoreService.candles(currency)

    override fun cryptoBalance(network: Network, currency: Currency): BigInt =
        walletService.balance(network, currency)

    override fun fiatBalance(network: Network, currency: Currency): Double {
        val mul = market(currency)?.currentPrice ?: 0.toDouble()
        val amount = cryptoBalance(network, currency)
        var balance = Formater.crypto(amount, currency.decimals(), mul)
        balance *= 100
        return balance.toInt().toDouble() / 100
    }

    override fun fiatPrice(amount: BigInt, currency: Currency): Double {
        val mul = market(currency)?.currentPrice ?: 0.toDouble()
        var price = Formater.crypto(amount, currency.decimals(), mul)
        price *= 100
        return price.toInt().toDouble() / 100
    }

    override fun transactions(network: Network, currency: Currency): List<AccountTransaction> =
        if (currency.type() == ERC20) {
            walletService.transferLogs(currency, network).toAccountTransactions(network)
        } else if (currency.isNative()) {
            etherScanService.transactionHistory(
                address(network),
                network,
                walletService.transactionCount(network).toDecimalString()
            ).toAccountTransactions()
        } else {
            emptyList()
        }


    override suspend fun fetchTransactions(network: Network, currency: Currency) {
        if (currency.type() == ERC20) {
            walletService.fetchTransferLogs(currency, network)
        } else if (currency.isNative()) {
            etherScanService.fetchTransactionHistory(address(network), network)
        }
    }

    override fun isVoidSigner(): Boolean =
        walletService.isSelectedVoidSigner()
}

private val Network.isEthereum: Boolean get() = chainId == Network.ethereum().chainId

private fun List<EtherScanTransaction>.toAccountTransactions(): List<AccountTransaction> =
    map {
        AccountTransaction(
            it.timeStamp.toIntOrNull(),
            it.blockNumber,
            it.from,
            it.to,
            BigInt.Companion.from(it.value),
            it.hash,
        )
    }.sortedByDescending { it.date }

private fun List<Log>.toAccountTransactions(network: Network): List<AccountTransaction> =
    mapNotNull {
        val topic1 = it.topics?.getOrNull(1)?.topicValueString
        val topic2 = it.topics?.getOrNull(2)?.topicValueString
        if (topic1 != null && topic2 != null) {
            AccountTransaction(
                BlockNumberToTimestampHelper.timestamp(it.blockNumber, network),
                it.blockNumber.toDecimalString(),
                abiDecodeAddress(topic1).hexString,
                abiDecodeAddress(topic2).hexString,
                abiDecodeBigInt(it.data),
                it.transactionHash,
            )
        } else { null }
    }.sortedByDescending { it.date }

private val Topic.topicValueString: String? get() =
    when(val topic = this) {
        is Topic.TopicValue -> topic.value.toString()
        else -> null
    }
